app/admin/payments.rb

```diff
@@ -5,14 +5,20 @@
 
     def edit
       @payment = Payment.find(permitted_params[:id])
+      unless @payment.refund
+        render :edit
+      else
+        flash[:notice] = "The product has already been refunded"
+        render :show
+      end
     end
 
     def update
       @payment = Payment.find(permitted_params[:id])
 
       respond_to do |format|
-        if @payment.update(permitted_params[:payment])
-          @payment.refund_functionality
+        if @payment.refund_functionality.status == 200
```

Comment:
`@payment.refund_functionality.status == 200` this seems like a good candidate for validation condition.

in controller you should just call `@payment.update` 
but in model validation you should check if there is refund present and what is the response of refund